index.js

```diff
@@ -543,147 +543,158 @@ async function html(urls, options) {
  */
 
 async function epubgen(data, output_path, options) {
-	const template_base = path.join(__dirname, 'templates/epub/');
-
-	const output = _fs.createWriteStream(output_path);
-	const archive = archiver('zip', {
-		store: true
-	});
-
-	output
-		.on('close', () => {
-			out.write(archive.pointer() + ' total bytes');
-			out.write(
-				'archiver has been finalized and the output file descriptor has closed.'
-			);
-		})
-		.on('end', () => {
-			out.write('Data has been drained');
+	const wrapAsync = inner => {
+		return new Promise((resolve, reject) => {
+			inner(resolve, reject).catch(reject);
 		});
+	};
 
-	archive
-		.on('warning', err => {
-			throw err;
-		})
-		.on('error', err => {
-			throw err;
-		})
-		.pipe(output);
+	return wrapAsync(async resolve => {
+		const template_base = path.join(__dirname, 'templates/epub/');
 
-	// mimetype file must be first
-	archive.append('application/epub+zip', { name: 'mimetype' });
-
-	// static files from META-INF
-	archive.directory(path.join(template_base, 'META-INF'), 'META-INF');
+		const output = _fs.createWriteStream(output_path);
+		const archive = archiver('zip', {
+			store: true
+		});
 
-	const contentTemplate = await fs.readFile(
-		path.join(template_base, 'OEBPS/content.xhtml'),
-		'utf8'
-	);
-	const navTemplate = await fs.readFile(
-		path.join(template_base, 'OEBPS/nav.xhtml'),
-		'utf8'
-	);
-	const tocTemplate = await fs.readFile(
-		path.join(template_base, 'OEBPS/toc.ncx'),
-		'utf8'
-	);
-	const opfTemplate = await fs.readFile(
-		path.join(template_base, 'OEBPS/content.opf'),
-		'utf8'
-	);
+		output
+			.on('close', () => {
+				out.write(archive.pointer() + ' total bytes');
+				out.write(
+					'archiver has been finalized and the output file descriptor has closed.'
+				);
+			})
+			.on('end', () => {
+				out.write('Data has been drained');
+			})
+			.on('finish', resolve);
+
+		archive
+			.on('warning', err => {
+				throw err;
+			})
+			.on('error', err => {
+				throw err;
+			})
```

Comment:
This should be fixed too, this will also lead to unhandled exceptions. If we go with the initial approach, this could be replaced by `reject(err)`.

Reply:
Sounds good!